### PR TITLE
Task5

### PR DESCRIPTION
**Pull request by moiroca**

### CODE REVIEW

File: webscraping/scraping.php [hunk 1]
```diff
@@ -0,0 +1,42 @@
+<?php
+  
+    
+    $username = "cl.epsteij11";
```
Comment (klagonoy) [inline webscraping/scraping.php]: Please stop committing credentials

File: webscraping/testHTMLScraper.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+$username = "admin";
```
Comment (klagonoy) [inline webscraping/testHTMLScraper.php]: Same here

File: webscraping/testScraper.php [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<html>
```
Comment (klagonoy) [inline webscraping/testScraper.php]: We don't need this for task5. Remove it

File: webscraping/scraping.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
+<?php
```
Comment (klagonoy) [inline webscraping/scraping.php]: We don't need the scraping files here. Remove everything related to web scraping

File: datastore/DataStore.php [hunk 5]
```diff
@@ -0,0 +1,255 @@
...
+        return (!empty($data) && isset($data[0]))?$this->cancelSubscription($data[0]["Id"]):null;
+    }
+
+    /**
+     * @param Integer $contact_id
+     */
+    public function getNumberOfSubscriptions($contact_id)
+    {
+       $query = array('ContactId' => (int)$contact_id);
+       $count = $isdk->dsCount("RecurringOrder",$query);
```
Comment (klagonoy) [inline datastore/DataStore.php]: Are you sure this is the right function the API uses? Please read the Infusion API documentation.

File: datastore/DataStore.php [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+    /**
+     * @param Email $email
+     */
+    public function cancelSubscriptionByEmail($email)
+    {
+        //Get All ContactId using email
+        $returnFields = array('Id');
+        $data = $isdk->findByEmail($email, $returnFields);
+        
+        return (!empty($data) && isset($data[0]))?$this->cancelSubscription($data[0]["Id"]):null;
```
Comment (klagonoy) [inline datastore/DataStore.php]: Where is cancelSubscription?

File: datastore/DataStore.php [hunk 7]
```diff
@@ -0,0 +1,255 @@
...
+    }
+
+    /**
+     * @param Email $email
+     */
+    public function getNumberOfSubscriptionsByEmail($email)
+    {
+        //get contact Id By Email
+        $returnFields = array('Id');
+        $data = $isdk->findByEmail($email, $returnFields);
```
Comment (klagonoy) [inline datastore/DataStore.php]: Please read the Infusion API documentation

File: datastore/DataStore.php [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+		if(count($invoice) == 0) return false;
+		
+		$invoice_id = (int)$invoice[0]['Id'];
+		
+		return $invoice[0]['PayStatus'] == 1 ? 'paid' : 'unpaid';
+	
+	}
+	
+    /**
+     * @param Integer $product_id
```
Comment (klagonoy) [inline datastore/DataStore.php]: $contact_id

File: datastore/DataStore.php [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+     */
+    public function cancelSubscription($contact_id)
+    {
+        $return_fields = array('Id');
+        $query = array('ContactId' => (int)$contact_id);
```
Comment (klagonoy) [inline datastore/DataStore.php]: Why do you need to force it to be int?

File: datastore/DataStore.php [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+        //get all subscription using contactId
+        $contacts = $isdk->dsQuery("RecurringOrder",null,null,$query,$return_fields);
+        
+        //loop through all subscription
+        foreach($contacts as $key=> $subscription)
```
Comment (klagonoy) [inline datastore/DataStore.php]: Is there a better way than this? I feel like you can do something like:
DELETE FROM <TABLE> WHERE subscription_id IN ($subscription_ids)

File: datastore/DataStore.php (hunk 9 above)
Comment (moiroca) [inline datastore/DataStore.php]: The user of this API might input an non numeric data. for security. I cast the contact id to int.

File: datastore/DataStore.php (hunk 8 above)
Comment (moiroca) [inline datastore/DataStore.php]: fixed!

File: datastore/DataStore.php (hunk 10 above)
Comment (moiroca) [inline datastore/DataStore.php]: I'll look back into the documentation if there is a deletion by group of ids.

File: datastore/DataStore.php [hunk 11]
```diff
@@ -0,0 +1,278 @@
+<?php
```
Comment (klagonoy) [inline datastore/DataStore.php]: Move this to the infusion folder

File: datastore/DataStore.php (hunk 9 above)
Comment (klagonoy) [inline datastore/DataStore.php]: Then do public function cancelSubscription(Integer $contact_id)

File: infusion/DataStore.php [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+	
+	}
+	
+    /**
+     * @param Integer $contact_id
+     */
+    public function cancelSubscription(Integer $contact_id)
+    {
+        $return_fields = array('Id');
+        $query = array('ContactId' => (int)$contact_id);
```
Comment (klagonoy) [inline infusion/DataStore.php]: Remove (int)

File: infusion/DataStore.php [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+    }
+    
+    /**
+     * @param Email $email
+     */
+    public function cancelSubscriptionByEmail($email)
+    {
+        //Get All ContactId using email
+        $returnFields = array('Id');
```
Comment (klagonoy) [inline infusion/DataStore.php]: Do not camel case variables

File: infusion/DataStore.php [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+
+		// close and download pdf report
+		$pdf->Output($file_name, 'I');
+    }    
+
+    public function getData($table_name, $display_per_page, $page, Array $fields, Array $where)
+    {
+       return $isdk->dsQuery($table_name, $display_per_page, $page, $fields,$where)
+    }
+}
```
Comment (klagonoy) [inline infusion/DataStore.php]: Add closing PHP tag

File: infusion/ReportBuilder.php [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+           $this->Cell($w, 6,$content[$header_index] , 'LR', 0, 'L', $fill);
+        }
+        
+        $this->Ln();
+        $fill=!$fill;
+      }
+                 
+      $this->Cell($w * count($headers), 0, '', 'T');
+  }
+}
```
Comment (klagonoy) [inline infusion/ReportBuilder.php]: Add closing PHP tag

File: infusion/ReportBuilder.php [hunk 16]
```diff
@@ -0,0 +1,43 @@
+<?php
+require_once('tcpdf_include.php');
+require_once('tcpdf.php');
+// extend TCPF with custom functions
+class ReportBuilder extends TCPDF {
+
+  // Colored table
+  public function formatReport($headers,$query) {
```
Comment (klagonoy) [inline infusion/ReportBuilder.php]: $data, not $query

File: infusion/ReportBuilder.php [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+           $this->Cell($width_size, 6,$content[$header_index] , 'LR', 0, 'L', $fill);
+        }
+        
+        $this->Ln();
+        $fill=!$fill;
+      }
+                 
+      $this->Cell($width_size * count($headers), 0, '', 'T');
+  }
+}
```
Comment (klagonoy) [inline infusion/ReportBuilder.php]: Closing PHP tag here

File: infusion/DataStore.php [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+
+		// close and output PDF document
+		$pdf->Output($file_name.".pdf", 'I');
+    }    
+
+    public function getData($table_name, $display_per_page, $page, Array $fields, Array $where)
+    {
+       return $isdk->dsQuery($table_name, $display_per_page, $page, $fields,$where);
+    }
+}
```
Comment (klagonoy) [inline infusion/DataStore.php]: Closing php tag here

File: infusion/DataStore.php [hunk 19]
```diff
@@ -0,0 +1,335 @@
...
+        return (!empty($data) && isset($data[0]))?$this->getNumberOfSubscriptions($data[0]["Id"]):null;
+    }
+
+    /*
+    *@return Download Reports in pdf format
+    */
+    public function downloadReports()
+    {
+
+    	include "ReportBuilder.php";
```
Comment (klagonoy) [inline infusion/DataStore.php]: Place this on the top of the file

**Pull request merged**